File: cohesiv/models/access_statement.ts
import * as Sequelize from 'sequelize'
import sequelize from '../db'

const AccessStatement = sequelize.define(
  'AccessStatement',
  {
    Sid: { type: Sequelize.STRING, primaryKey: true, field: 'sid' },
    Effect: {
      type: Sequelize.STRING,
      allowNull: false,
      field: 'effect'
    },
    Action: {
      type: Sequelize.ARRAY(Sequelize.TEXT),
      allowNull: false,
      field: 'action'
    }, // Defines an array. PostgreSQL only.
    Resource: {
      type: Sequelize.ARRAY(Sequelize.TEXT),
      allowNull: false,
      field: 'resource'
    } // Defines an array. PostgreSQL only.
  },
  {
    tableName: 'access_statements',
    underscored: true
  }
)

AccessStatement.sync({
  force: process.env.NODE_ENV === 'production' ? false : true
})

export default AccessStatement

File: cohesiv/models/access_statement.go
package models

import (
	"time"

	pb "github.com/SingularityInteractive/cohesiv/cohesiv"
	"github.com/lib/pq"
)

// AccessStatementModel as represented in database
type AccessStatementModel struct {
	Sid       string         `gorm:"primary_key;column:sid"`
	Effect    string         `gorm:"column:effect"`
	Action    pq.StringArray `gorm:"column:action;type:varchar(64)[]"`
	Resource  pq.StringArray `gorm:"column:resource;type:varchar(64)[]"`
	CreatedAt time.Time
	UpdatedAt time.Time
}

// TableName renames the table on migration with gorm
func (AccessStatementModel) TableName() string {
	return "access_statements"
}

// ToProto converts a db access statement to grpc access statement
func (a *AccessStatementModel) ToProto() *pb.AccessStatement {
	return &pb.AccessStatement{
		Sid:      a.Sid,
		Effect:   a.Effect,
		Action:   a.Action,
		Resource: a.Resource,
	}
}
